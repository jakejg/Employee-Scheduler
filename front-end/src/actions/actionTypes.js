export const LOAD_JOBS = "LOAD_JOBS";
export const ADD_JOBS = "ADD_JOB";