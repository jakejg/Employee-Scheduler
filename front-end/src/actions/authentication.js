import axios from 'axios';
import { 
    ADD_OR_REMOVE_TOKEN
 } from './actionTypes';
import {addCompanyToAPI} from '../helpers/JobApi'
import {BASE_URL} from '../config';


export const register = (data) => {
    
    return async (dispatch) => {
        try {
            // designate admin user
            data.is_admin = true
            // create admin user
       
            const res = await axios.post(`${BASE_URL}/auth/register`, data);
       
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const login = (data) => {
    return async (dispatch) => {
        try {
            let res = await axios.post(`${BASE_URL}/auth/login`, data);
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            // console.log(e.response.data.message)
            return e.response.data.message
        }
    }
}

export const addOrRemoveToken = (token) => {
    return {type: ADD_OR_REMOVE_TOKEN, token}
}
