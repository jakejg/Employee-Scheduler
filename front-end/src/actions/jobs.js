import { 
    LOAD_JOBS,
    EDIT_JOB, 
    ADD_JOB,
    EDIT_JOB_STAFF,
    DELETE_JOB
 } from './actionTypes';
import {JobAPI} from '../helpers/JobApi';

export const loadJobsFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            const jobData = await JobAPI.loadJobs(comp_id)
            dispatch(loadJobs(jobData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadJobs = (jobs) => {
    return {type: LOAD_JOBS, jobs}
}

export const addJobOnAPI = (jobToAdd) => {

    return async (dispatch) => {
        try {
            let job = await JobAPI.addJob(jobToAdd)
            dispatch(addJob(job.id, job))
            return {message: "Job created Successfully", severity: "success"}
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addJob = (id, job) => {
    return {type: ADD_JOB, id, job}
}


export const getJobFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.getJob(ID)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJobOnAPI = (id, changedProps) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.editJob(id, changedProps)
            console.log(job)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJob = (id, job) => {
    return {type: EDIT_JOB, id, job}
}

export const addStaffToJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.addStaffToJob(jobId, staffId)
            console.log(data)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const removeStaffFromJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.removeStaffFromJob(jobId, staffId)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const editJobStaff = (id, staff, staff_filled) => {
    return {type: EDIT_JOB_STAFF, id, staff, staff_filled}
}

export const deleteJobOnAPI = (id) => {
    return async (dispatch) => {
        try {
            let msg = await JobAPI.deleteJob(id)
            dispatch(deleteJob(id))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const deleteJob = (id) => {
    return {type: DELETE_JOB, id}
}