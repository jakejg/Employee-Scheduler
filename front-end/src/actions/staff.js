import { 
    LOAD_STAFF,
    EDIT_STAFF,
    ADD_STAFF
 } from './actionTypes';
import { StaffAPI } from '../helpers/StaffApi';

export const loadStaffFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            let staffData = await StaffAPI.loadStaff(comp_id);
            dispatch(loadStaff(staffData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadStaff = (staff) => {
    return {type: LOAD_STAFF, staff}
}


export const getStaffFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let staff = await StaffAPI.getStaff(ID)
        
            dispatch(editStaff(staff.id, staff))
            }
        catch(e) {
            console.log(e)
        }
    }
}

export const editStaff = (id, staff) => {
    return {type: EDIT_STAFF, id, staff}
}

export const addStaffOnAPI = (staffToAdd) => {
   
    return async (dispatch) => {
        try {
            const staff = await StaffAPI.addStaff(staffToAdd) ;
            dispatch(addStaff(staff.id, staff))
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addStaff = (id, staff) => {
    return {type: ADD_STAFF, id, staff}
}