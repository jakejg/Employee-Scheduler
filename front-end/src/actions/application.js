import {CHANGE_DRAWER, CHANGE_NO_JOBS, CHANGE_NO_STAFF} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}

export const changeNoJobs = () => {
    return {type: CHANGE_NO_JOBS}
}

export const changeNoStaff = () => {
    return {type: CHANGE_NO_STAFF}
}