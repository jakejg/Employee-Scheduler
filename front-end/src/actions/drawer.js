import {CHANGE_DRAWER} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}