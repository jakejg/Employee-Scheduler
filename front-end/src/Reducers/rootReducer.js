import jobs from './jobsReducer';
import staff from './staffReducer';
import { combineReducers } from "redux";

export default combineReducers({jobs, staff})