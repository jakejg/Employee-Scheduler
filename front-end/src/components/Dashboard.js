import React, {useEffect} from 'react';
import clsx from 'clsx';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { useSelector, useDispatch } from 'react-redux';
import ItemList from './ItemList';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Sidebar from './Sidebar';
import JobList from './JobList';
import StaffList from './StaffList';
import { loadJobsFromAPI } from '../actions/jobs';
import { loadStaffFromAPI } from '../actions/staff';
import { decode } from "jsonwebtoken";

const drawerWidth = 240;

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
  },

  content: {
    flexGrow: 1,
    padding: theme.spacing(3),
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    marginLeft: -drawerWidth,
  },
  contentShift: {
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginLeft: 0,
  },
}));

const Dashboard = () => {
    const classes = useStyles();
    const theme = useTheme();
    const open = useSelector(state => state.application.drawer)
    const token = useSelector(state => state.application.token)
    const dispatch = useDispatch();

    useEffect(() => {
      const getData = async () => {
          const { comp_id } = decode(token);
          dispatch(loadJobsFromAPI(comp_id))
          dispatch(loadStaffFromAPI(comp_id))
      }
      getData();
  }, [dispatch])

    const jobFields = ['Title', 'Start Date','End Date', 'Staff Needed', 'Notes'];
    const staffFields = ['Username', 'First Name','Last Name', 'Current Wage', 'Years At Company']
    return (
      <div className={classes.root}>
            <Sidebar jobFields={jobFields} staffFields={staffFields}/>
        <div className={clsx(classes.content, {
            [classes.contentShift]: open,
          })} >
              <Container>
                    <h1 style={{textAlign: 'center'}}>Dashboard</h1>
                    <Grid container spacing={3} className={classes.drawerHeader} >
                         <Grid item xs>
                             <Paper>
                                <JobList />
                             </Paper>
                         </Grid>
                         <Grid item xs>
                             <Paper>
                                <StaffList />
                             </Paper>
                         </Grid>

                    </Grid>
               </Container>
        </div>
      
      </div>
    );
}

export default Dashboard;
