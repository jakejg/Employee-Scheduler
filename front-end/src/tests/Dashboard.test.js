import React from 'react';
import { render } from '@testing-library/react';
import {Provider} from 'react-redux'
import Dashboard from '../components/Dashboard';
import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/rootReducer";
import { MemoryRouter } from "react-router";
import { addOrRemoveToken } from '../actions/authentication';
import thunk from 'redux-thunk'

const store = createStore( rootReducer, applyMiddleware(thunk));
store.dispatch(addOrRemoveToken("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6IkRlbW8iLCJpc19hZG1pbiI6dHJ1ZSwiY29tcF9pZCI6MSwiaWF0IjoxNTk3NzY5NTYyfQ.Wm2z3GCj-m6DDWElCIBCuRxqBbc3_0QhJzeorsbz6E0")) 

it('renders without crashing', () => {

  render(
    <MemoryRouter>
      <Provider store={store}>
      <Dashboard />
      </Provider>
    </MemoryRouter>
)
});

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Provider store={store}>
                <Dashboard />
            </Provider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})