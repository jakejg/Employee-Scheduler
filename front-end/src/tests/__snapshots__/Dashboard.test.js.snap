// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-24"
  >
    <div
      class="makeStyles-root-27"
    >
      <div
        class="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-32"
      >
        <div
          class="MuiPaper-root MuiDrawer-paper makeStyles-drawerPaper-33 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
          style="visibility: hidden; webkit-transform: translateX(-0px); transform: translateX(-0px);"
        >
          <div
            class="makeStyles-drawerHeader-34"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <hr
            class="MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 6h-4V4c0-1.11-.89-2-2-2h-4c-1.11 0-2 .89-2 2v2H4c-1.11 0-1.99.89-1.99 2L2 19c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V8c0-1.11-.89-2-2-2zm-6 0h-4V4h4v2z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Add Job
                </span>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm-9-2V7H4v3H1v2h3v3h2v-3h3v-2H6zm9 4c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Add Staff
                </span>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 10H7v2h10v-2zm2-7h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zm-5-5H7v2h7v-2z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Check Dates
                </span>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </ul>
          <hr
            class="MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                  />
                </svg>
              </div>
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  View Calendar
                </span>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="makeStyles-content-25"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiBox-root MuiBox-root-37"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter"
            >
              Dashboard
            </h3>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
            >
              <div
                class="MuiBox-root MuiBox-root-43"
              >
                <div
                  class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </div>
                <h3
                  class="MuiTypography-root MuiTypography-h3"
                >
                  Loading
                </h3>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
            >
              <div
                class="MuiBox-root MuiBox-root-46"
              >
                <div
                  class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </div>
                <h3
                  class="MuiTypography-root MuiTypography-h3"
                >
                  Loading
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
